tower-http/src/decompression.rs

```diff
@@ -0,0 +1,692 @@
+    pub fn no_br(self) -> Self {
+        self.accept.set_br(false);
+        self
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for Decompress<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Body,
+{
+    type Response = Response<DecompressBody<ResBody>>;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        if !req.headers().contains_key(RANGE) {
+            if let header::Entry::Vacant(e) = req.headers_mut().entry(ACCEPT_ENCODING) {
+                if let Some(accept) = self.accept.to_header_value() {
+                    e.insert(accept);
+                }
+            }
+        }
+        ResponseFuture {
+            inner: self.inner.call(req),
+            accept: self.accept,
+        }
+    }
+}
+
+impl DecompressLayer {
+    /// Creates a new `DecompressLayer`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Sets whether to request the gzip encoding.
+    #[cfg(feature = "decompression-gzip")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-gzip")))]
+    pub fn gzip(self, enable: bool) -> Self {
+        self.accept.set_gzip(enable);
+        self
+    }
+
+    /// Sets whether to request the Deflate encoding.
+    #[cfg(feature = "decompression-deflate")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-deflate")))]
+    pub fn deflate(self, enable: bool) -> Self {
+        self.accept.set_deflate(enable);
+        self
+    }
+
+    /// Sets whether to request the Brotli encoding.
+    #[cfg(feature = "decompression-br")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "decompression-br")))]
+    pub fn br(self, enable: bool) -> Self {
+        self.accept.set_br(enable);
+        self
+    }
+
+    /// Disables the gzip encoding.
+    ///
+    /// This method is available even if the `gzip` crate feature is disabled.
+    pub fn no_gzip(self) -> Self {
+        self.accept.set_gzip(false);
+        self
+    }
+
+    /// Disables the Deflate encoding.
+    ///
+    /// This method is available even if the `deflate` crate feature is disabled.
+    pub fn no_deflate(self) -> Self {
+        self.accept.set_deflate(false);
+        self
+    }
+
+    /// Disables the Brotli encoding.
+    ///
+    /// This method is available even if the `br` crate feature is disabled.
+    pub fn no_br(self) -> Self {
+        self.accept.set_br(false);
+        self
+    }
+}
+
+impl<S> Layer<S> for DecompressLayer {
+    type Service = Decompress<S>;
+
+    fn layer(&self, service: S) -> Self::Service {
+        Decompress {
+            inner: service,
+            accept: self.accept,
+        }
+    }
+}
+
+impl<F, B> Future for ResponseFuture<F>
+where
+    F: TryFuture<Ok = Response<B>>,
```

Comment:
I think we should just use `std::future::Future` here instead of `TryFuture`. Would like to avoid referencing `futures` things in public APIs since then we're depending on them not making breaking changes. I will add a few more generic params but I'm not too worried about that.